### PR TITLE
Fix importing string sets

### PR DESCRIPTION
**Pull request by danielrbradley**
Fixes https://github.com/pulumi/pulumi-azure-native/issues/2933

The method `convertPropValue` is currently used during three distinct operations:
1. Converting SDK inputs into an Azure request body, where string sets should be changed from string arrays to maps.
2. Converting an Azure response body into SDK outputs, where there should be no change to string arrays.
3. Converting SDK outputs back into SDK inputs, where string sets should be changed from maps into string arrays.

The [original String Set implementation](https://github.com/pulumi/pulumi-azure-native/pull/2524) was incorrect because it only dealt with (1) but not (3). However, if we fix the behaviour for (3) then we'll break the implementation for (2). 

Therefore this PR creates 3 separate copies of `convertPropValue` - each one specialized for each operation. 

When in an "import" operation, there's a special code path due to not having old inputs. On this code path we only converted the Azure response body back to outputs, but did not convert them back to inputs which was incorrect. This was caused by poor naming of the methods.

In addition, to make the converter easier to comprehend we:
1. Move the converter into its own module so we can be confident about how it's interacted with.
2. Split each public converter method into its own file with all associated code alongside.
3. Rename and document methods to accurately describe their purpose.
4. Rewrite the tests to be oriented around the public entry point with finely grained cases testing a specific expected behaviour.

## Detailed breakdown

1. Identify all callers of `convertPropValue` which provide the rest of the implementation via a callback argument.
5. Create named versions of `convertPropValue` for each of possible configured callbacks:
    1. `convertBodyPropToSdkPropValue`
    5. `convertOutputToInputPropValue`
    6. `convertSdkPropToRequestBodyPropValue`
6. Move implementation and tests for `convert.go` into its own module so we can be sure of the external interface before we write more tests.
7. Split implementation of `convert.go` into a source file per entry point:
    1. BodyPropertiesToSdk
    2. IsDefaultResponse
    3. PreviewOutputs
    7. ResponseToSdkInputs
    8. SdkOutputsToSdkInputs
    9. NewSdkShapeConverter
    10. SdkToRequestBody
8. Co-locate tests in file alongside the corresponding entry point that they're testing for.
9. Re-write tests to test a specific feature per test rather than a single massive input that attempts (and fails) to test all edge cases.

<details>
<summary>Generated output</summary>

```typescript

import * as pulumi from "@pulumi/pulumi";
import * as azure_native from "@pulumi/azure-native";

const ringoimportapigw = new azure_native.network.ApplicationGateway("ringoimportapigw", {
    applicationGatewayName: "ringoimportapigwed0a4e0e",
    backendAddressPools: [{
        id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/backendAddressPools/defaultBackendAddressPool",
        name: "defaultBackendAddressPool",
    }],
    backendHttpSettingsCollection: [{
        cookieBasedAffinity: "Disabled",
        id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/backendHttpSettingsCollection/defaultBackendHttpSettings",
        name: "defaultBackendHttpSettings",
        pickHostNameFromBackendAddress: false,
        port: 40000,
        protocol: "Http",
        requestTimeout: 30,
    }],
    frontendIPConfigurations: [{
        id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/frontendIPConfigurations/myFrontendIp",
        name: "myFrontendIp",
        privateIPAllocationMethod: "Dynamic",
        publicIPAddress: {
            id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/publicIPAddresses/ringoimportapigw3fb87c0f",
        },
    }],
    frontendPorts: [{
        id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/frontendPorts/port80",
        name: "port80",
        port: 80,
    }],
    gatewayIPConfigurations: [{
        id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/gatewayIPConfigurations/myGatewayIpConfig",
        name: "myGatewayIpConfig",
        subnet: {
            id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/virtualNetworks/ringoimportapigw5b165c75/subnets/subnet1",
        },
    }],
    httpListeners: [{
        frontendIPConfiguration: {
            id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/frontendIPConfigurations/myFrontendIp",
        },
        frontendPort: {
            id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/frontendPorts/port80",
        },
        id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/httpListeners/myListener",
        name: "myListener",
        protocol: "Http",
        requireServerNameIndication: false,
    }],
    identity: {
        type: azure_native.network.ResourceIdentityType["userAssigned"],
        userAssignedIdentities: ["/subscriptions/<sub-id>/resourcegroups/ringoimportapigweac2fabe/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ringoimportapigw10cfe6d3"],
    },
    location: "westeurope",
    requestRoutingRules: [{
        backendAddressPool: {
            id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/backendAddressPools/defaultBackendAddressPool",
        },
        backendHttpSettings: {
            id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/backendHttpSettingsCollection/defaultBackendHttpSettings",
        },
        httpListener: {
            id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/httpListeners/myListener",
        },
        id: "/subscriptions/<sub-id>/resourceGroups/ringoimportapigweac2fabe/providers/Microsoft.Network/applicationGateways/ringoimportapigwed0a4e0e/requestRoutingRules/rule1",
        name: "rule1",
        priority: 1,
        ruleType: "Basic",
    }],
    resourceGroupName: "ringoimportapigweac2fabe",
    sku: {
        capacity: 2,
        name: "Standard_v2",
        tier: "Standard_v2",
    },
}, {
    protect: true,
});
```
</details>

### CODE REVIEW

File: provider/pkg/resources/convert.go [hunk 1]
```diff
@@ -93,7 +93,15 @@ func (k *SdkShapeConverter) convertPropValue(prop *AzureAPIProperty, value inter
 			}
 			return result
 		}
-
+		// Convert can be called for either turning user inputs into a request body or for turning a response body
+		// into SDK outputs. In the latter case, we need to turn string sets back into arrays.
```
Comment (thomas11) [inline provider/pkg/resources/convert.go:97]: The comment says "in the latter case" but how do we know what case we have here?

File: provider/pkg/resources/convert.go [hunk 2]
```diff
@@ -93,7 +93,15 @@ func (k *SdkShapeConverter) convertPropValue(prop *AzureAPIProperty, value inter
...
+		if prop.IsStringSet {
```
Comment (thomas11) [inline provider/pkg/resources/convert.go:98]: Since this method has two distinct use cases and several code paths, some tests would help assure that the change has no unintended consequences on other scenarios.

File: provider/pkg/resources/convert.go (hunk 1 above)
Comment (danielrbradley) [reply provider/pkg/resources/convert.go:97]: That's the fun part ... we don't. The same method is called for both directions with no indicator as to which way we're going. Definitely on my hit-list for testing thoroughly and rewriting.

File: provider/pkg/convert/sdkInputsToRequestBody.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/pulumi/pulumi-azure-native/v2/provider/pkg/resources"
+)
+
+// SdkInputsToRequestBody converts a map of SDK properties to JSON request body to be sent to an HTTP API.
+func (k *SdkShapeConverter) SdkInputsToRequestBody(props map[string]resources.AzureAPIProperty,
+	values map[string]interface{}, id string) map[string]interface{} {
+	result := map[string]interface{}{}
+
+	for name, prop := range props {
+		p := prop // https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
```
Comment (thomas11) [inline provider/pkg/convert/sdkInputsToRequestBody.go:18]: [nit] this url has moved to https://go.dev/wiki/CommonMistakes#using-reference-to-loop-iterator-variable

**Review by thomas11 — COMMENTED** [review]
Approving with a few non-essential suggestions. Awesome test coverage!

**Review by thomas11 — APPROVED** [review]

File: provider/pkg/convert/previewOutputs.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		result := resource.PropertyMap{}
+		for name, value := range inputObject {
+			p := value
+			result[name] = k.previewOutputValue(p, prop.AdditionalProperties)
+		}
+		return resource.NewObjectProperty(result)
+
+	case prop.OneOf != nil && inputValue.IsObject():
+		// TODO: It would be nice to do something smart here and pick the right oneOf branch based on the input.
+		// The challenge is differentiating between a mis-match and valid unknowns in the input.
```
[the remark this answers is not in the document]
Comment (thomas11) [reply provider/pkg/convert/previewOutputs.go:76]: (the code especially)

File: provider/pkg/convert/isDefaultResponse.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+//     non-empty collections, but none are found yet)
+func (k *SdkShapeConverter) IsDefaultResponse(putParameters []resources.AzureAPIParameter, response map[string]interface{},
+	defaultBody map[string]interface{}) bool {
+	for _, param := range putParameters {
+		if param.Location == body {
+			for key, value := range k.ResponseBodyToSdkOutputs(param.Body.Properties, response) {
+				switch reflect.TypeOf(value).Kind() {
+				case reflect.Slice, reflect.Array:
+					collection := reflect.ValueOf(value)
+					if collection.Len() > 0 {
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/convert/isDefaultResponse.go:23]: Yeah, apparently we don't support checking for this right now. Looking at our [default resource states](https://github.com/pulumi/pulumi-azure-native/blob/44e1efe11d5e920caf0c47392e7670fa60e64cec/provider/pkg/openapi/defaults/defaultResourcesState.go), there's no use of arrays yet. Will add a comment.

File: provider/pkg/convert/convertTypedObjects.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2016-2020, Pulumi Corporation.
+
+package convert
+
+import (
+	"github.com/pulumi/pulumi-azure-native/v2/provider/pkg/resources"
+)
+
+// Returning nil indicates that the given value is not of the expected type.
+type convertTypedObject func(typeName string, props map[string]resources.AzureAPIProperty, values map[string]interface{}) map[string]interface{}
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/convert/convertTypedObjects.go:10]: It's any arbitary conversion hook - depending on where it's being used. Added a comment to indicate this.

File: provider/pkg/convert/previewOutputs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/convert/previewOutputs.go:76]: Yeah, I think this is still a valid theoretical question but I don't think we need to implement it yet either.
